README.md
# Koop-Worker
* A worker bot to run jobs for Koop

[![npm][npm-img]][npm-url]
[![travis][travis-image]][travis-url]

## Usage
- `npm install koop-worker`
- edit configs
- `require('koop-worker')`

[npm-img]: https://img.shields.io/npm/v/@koopjs/worker.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@koopjs/worker
[travis-image]: https://img.shields.io/travis/koopjs/koop-worker.svg?style=flat-square
[travis-url]: https://travis-ci.org/koopjs/koop-worker


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.3.5] - 2017-04-13
### Fixed
* Catch error in preparing query during export

## [1.3.4] - 2017-04-13
### Changed
* Upgrade required version of @koopjs/filesystem-s3

## [1.3.3] - 2017-03-27
### Changed
* Support Winnow that has outSR

## [1.3.2] - 2017-01-31
### Fixed
* Recompiled

## [1.3.1] - 2017-01-31
### Changed
* Share code for initializing Koop between `copyFile` and `exportFile`

## [1.3.0] - 2017-01-31
### Changed
* Compatible with Koop 3.x latest

## [1.2.10] - 2016-11-29
### Changed
* Delay 30 seconds before retrying a job

## [1.2.9] - 2016-10-29
### Fixed
* Ensure heartbeat is not sent after job completes

## [1.2.8] - 2016-10-15
### Changed
* Reduce geohash precision to 7

### Fixed
* Retrieval time is written to s3 metadata

## [1.2.7] - 2016-05-13
### Fixed
* Files are written into the correct temporary directory

## [1.2.6] - 2016-05-12
### Fixed
* Uncommented line causing multiple callbacks to fire at the end of export jobs

## [1.2.5] - 2016-05-06
### Changed
* Set the content type for files saved on s3

## [1.2.4] - 2016-05-06
### Changed
* Jobs running over 60 minutes are cancelled

### Fixed
* Catch ECONNRESET error from source in export stream

## [1.2.3] - 2016-04-28
### Fixed
* No unhandled rejections shall go uncaught
* Do not leave temp directory lying around when the source stream fails

## [1.2.2] - 2016-04-25
### Fixed
* Recommend retry when geohash fails because geojson is not ready on S3
* Handle geohash errors without throwing an exception

## [1.2.1] - 2016-04-22
### Changed
* Abort geoxform if write to disk fails in exportFile
* Shut down on emfile errors in exportFile

## [1.2.0] - 2016-04-21
### Added
* `maxRetries` is an new option for jobs

### Changed
* Only retry on some types of errors
* Remove `plugins/retry`

## [1.1.3] - 2016-04-14
### Changed
* Integrate koop-s3fs

## [1.1.2] - 2016-04-13
### Fixed
* Don't pass metadata object to S3 if the values are empty

## [1.1.1] - 2016-04-04
### Changed
* Metadata object is stored with files saved on S3

## [1.1.0] - 2016-03-22
### Added
* ExportFile can generate a geohash

## [1.0.1] - 2016-03-22
### Fixed
* Rebuild

## [1.0.0] - 2016-03-22
### Added
* Use Winnow for creating filtered files

### Changed
* ExportFile Job now takes a source (geojson filepath) and an output. The format, name and filePath parameters are no longer accepted

## [0.3.1] - 2016-03-17
### Changed
* Upgrade config package

## [0.3.0] - 2016-03-16
### Added
* Export jobs are retried 3 times

## [0.2.0] - 2016-03-15
### Added
* Clean shutdown on SIGTERM

## [0.1.1] - 2016-01-27
### Added
* Supply temp path in exportFile job

## [0.1.0] - 2016-01-26
### Changed
* `Rename xport to ExportFile`
* Remove simultaneous GeoJSON + transform writing
* Refactor exortFile internals

## [0.0.4] - 2016-01-26
### Changed
* Use fork instead of observe to prevent memory issues

## [0.0.3] - 2016-01-12
### Fixed
* Require Koop > 3.0

## [0.0.2] - 2016-01-08
### Fixed
* Removed case of trying to JSON.stringify circular object

## [0.0.1] - 2016-01-07
### Added
* First release

[1.3.5]: https://github.com/koopjs/koop-worker/compare/v1.3.4..v1.3.5
[1.3.4]: https://github.com/koopjs/koop-worker/compare/v1.3.4..v1.3.3
[1.3.3]: https://github.com/koopjs/koop-worker/compare/v1.3.2..v1.3.3
[1.3.2]: https://github.com/koopjs/koop-worker/compare/v1.3.2..v1.3.1
[1.3.1]: https://github.com/koopjs/koop-worker/compare/v1.3.1..v1.3.0
[1.3.0]: https://github.com/koopjs/koop-worker/compare/v1.3.0..v1.2.10
[1.2.10]: https://github.com/koopjs/koop-worker/compare/v1.2.10..v1.2.9
[1.2.9]: https://github.com/koopjs/koop-worker/compare/v1.2.9..v1.2.8
[1.2.8]: https://github.com/koopjs/koop-worker/compare/v1.2.8..v1.2.7
[1.2.7]: https://github.com/koopjs/koop-worker/compare/v1.2.7..v1.2.6
[1.2.6]: https://github.com/koopjs/koop-worker/compare/v1.2.6..v1.2.5
[1.2.5]: https://github.com/koopjs/koop-worker/compare/v1.2.5..v1.2.4
[1.2.4]: https://github.com/koopjs/koop-worker/compare/v1.2.4..v1.2.3
[1.2.3]: https://github.com/koopjs/koop-worker/compare/v1.2.3..v1.2.2
[1.2.2]: https://github.com/koopjs/koop-worker/compare/v1.2.2..v1.2.1
[1.2.1]: https://github.com/koopjs/koop-worker/compare/v1.2.1..v1.2.0
[1.2.0]: https://github.com/koopjs/koop-worker/compare/v1.2.0..v1.1.3
[1.1.3]: https://github.com/koopjs/koop-worker/compare/v1.1.3..v1.1.2
[1.1.2]: https://github.com/koopjs/koop-worker/compare/v1.1.2..v1.1.1
[1.1.1]: https://github.com/koopjs/koop-worker/compare/v1.1.1..v1.1.0
[1.1.0]: https://github.com/koopjs/koop-worker/compare/v1.1.0..v1.0.1
[1.0.1]: https://github.com/koopjs/koop-worker/compare/v1.0.0..v1.0.1
[1.0.0]: https://github.com/koopjs/koop-worker/compare/v0.3.1..v1.0.0
[0.3.1]: https://github.com/koopjs/koop-worker/compare/v0.3.0..v0.3.1
[0.3.0]: https://github.com/koopjs/koop-worker/compare/v0.2.0..v0.3.0
[0.2.0]: https://github.com/koopjs/koop-worker/compare/v0.1.1..v0.2.0
[0.1.1]: https://github.com/koopjs/koop-worker/compare/v0.1.0..v0.1.1
[0.1.0]: https://github.com/koopjs/koop-worker/compare/v0.0.4..v0.1.0
[0.0.4]: https://github.com/koopjs/koop-worker/compare/v0.0.3..v0.0.4
[0.0.3]: https://github.com/koopjs/koop-worker/compare/v0.0.2..v0.0.3
[0.0.2]: https://github.com/koopjs/koop-worker/compare/v0.0.1..v0.0.2
[0.0.1]: https://github.com/koopjs/koop-worker/tree/v0.0.1


test/index.js
/* @flow */
'use strict'
const test = require('tape')
const NR = require('node-resque')
const config = require('config')
const rimraf = require('rimraf')
const connection = config.queue.connection || {}
connection.pkg = 'redis'

const queue = new NR.queue({connection}) // eslint-disable-line
const Redis = require('redis')
const redis = Redis.createClient(connection)

const worker = require('../src')

test('Set up', t => {
  queue.connect(() => {
    queue.connection.redis.del(`${connection.namespace}:queue:koop`, e => {
      t.end()
    })
  })
})

test('Run an exportFile job that succeeds', t => {
  t.plan(4)
  const options = {
    id: 'f445febc447d4cb696e71ea7816d65d5',
    layer: 0,
    output: 'files/f445febc447d4cb696e71ea7816d65d5_0/full_0/test.csv',
    source: `files/f445febc447d4cb696e71ea7816d65d5_0/full_0/test.geojson`,
    job_id: 'job_id'
  }
  redis.subscribe('jobs', () => {
    queue.enqueue('koop', 'exportFile', options)
    redis.once('message', (channel, message) => {
      const info = JSON.parse(message)
      t.equal(info.status, 'start', 'Start emitted')
      t.equal(info.id, 'job_id', 'Job id matches')
      redis.once('message', (channel, message) => {
        const info = JSON.parse(message)
        t.equal(info.status, 'finish', 'Finish emitted')
        t.equal(info.id, 'job_id', 'Job id matches')
      })
    })
  })
})

test('Run an exportFile job that fails', t => {
  t.plan(4)
  const options = {
    id: 'f445febc447d4cb696e71ea7816d65d5',
    layer: 0,
    output: 'files/f445febc447d4cb696e71ea7816d65d5_0/full_0/test.csv',
    source: `files/f445febc447d4cb696e71ea7816d65d5_0/full_0/empty.geojson`,
    job_id: 'job_id2',
    maxRetries: 0
  }
  redis.subscribe('jobs', () => {
    queue.enqueue('koop', 'exportFile', options)
    redis.once('message', (channel, message) => {
      const info = JSON.parse(message)
      t.equal(info.status, 'start', 'Start emitted')
      t.equal(info.id, 'job_id2', 'Job id matches')
      redis.once('message', (channel, message) => {
        const info = JSON.parse(message)
        t.equal(info.status, 'fail', 'Fail emitted')
        t.equal(info.id, 'job_id2', 'Job id matches')
      })
    })
  })
})

test('Run an exportFile that fails, retries and fails again', t => {
  t.plan(8)
  const options = {
    id: 'f445febc447d4cb696e71ea7816d65d5',
    layer: 0,
    output: 'files/f445febc447d4cb696e71ea7816d65d5_0/full_0/test.csv',
    source: `files/f445febc447d4cb696e71ea7816d65d5_0/full_0/empty.geojson`,
    job_id: 'job_id3',
    maxRetries: 1
  }
  redis.subscribe('jobs', () => {
    queue.enqueue('koop', 'exportFile', options)
    redis.once('message', (channel, message) => {
      const info = JSON.parse(message)
      t.equal(info.status, 'start', 'Start emitted')
      t.equal(info.id, 'job_id3', 'Job id matches')
      redis.once('message', (channel, message) => {
        const info = JSON.parse(message)
        t.equal(info.status, 'retry', 'Retry emitted')
        t.equal(info.id, 'job_id3', 'Job id matches')
        redis.once('message', (channel, message) => {
          const info = JSON.parse(message)
          t.equal(info.status, 'start', 'Job restarted')
          t.equal(info.id, 'job_id3', 'Job id matches')
          redis.once('message', (channel, message) => {
            const info = JSON.parse(message)
            t.equal(info.status, 'fail', 'Job failed')
            t.equal(info.id, 'job_id3', 'Job id matches')
          })
        })
      })
    })
  })
})

test('Teardown', t => {
  worker.end()
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_0/test_0')
  rimraf.sync('./test/data/files/foo_0')
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_0/full_0/test.csv')
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_3')
  queue.end()
  redis.quit()
  t.end()
})


test/xform.js
const test = require('tape')
const rewire = require('rewire')
const xform = rewire('../src/jobs/xform')
const fs = require('fs')
const _ = require('highland')
const rimraf = require('rimraf')

test('Transform data when geojson already exists', t => {
  t.plan(2)
  const options = {
    id: 'f445febc447d4cb696e71ea7816d65d5',
    layer: 0,
    key: 'test',
    format: 'csv',
    name: 'test',
    path: './test/output'
  }
  xform(options, err => {
    t.error(err, 'No error')
    const rows = []
    _(fs.createReadStream(fileLoc(options)))
    .split()
    .compact()
    .each(r => rows.push(r))
    .done(() => t.equal(rows.length, 101), 'CSV has all expected features')
  })
})

test('Set up feature stream mock', t => {
  function createStream () {
    return _(fs.createReadStream('./test/fixtures/features.txt')).split().compact()
  }
  function exists (a, b, callback) { callback(false) }
  xform.__set__('koop.Cache.createStream', createStream)
  xform.__set__('koop.files.exists', exists)
  t.end()
})

test('Transform data and save geojson when geojson does not exist', t => {
  t.plan(3)
  const options = {
    id: 'test',
    layer: 0,
    key: 'test',
    format: 'csv',
    name: 'test',
    path: './test/output'
  }
  xform(options, err => {
    t.error(err, 'No error')

    const path = fileLoc(options)
    const rows = []
    _(fs.createReadStream(path))
    .split()
    .compact()
    .each(r => rows.push(r))
    .done(() => t.equal(rows.length, 101), 'CSV has all expected features')

    const geojson = fs.readFileSync(path.replace('csv', 'geojson'))
    const features = JSON.parse(geojson).features
    t.equal(features.length, 100, 'GeoJSON has all expected features')
  })
})

test('Transform data directly into geojson', t => {
  t.plan(2)
  const options = {
    id: 'test',
    layer: 1,
    key: 'test',
    format: 'geojson',
    name: 'test',
    path: './test/output'
  }
  xform(options, err => {
    t.error(err, 'No error')
    const path = fileLoc(options)
    const geojson = fs.readFileSync(path)
    const features = JSON.parse(geojson).features
    t.equal(features.length, 100, 'GeoJSON has all expected features')
  })
})

test('Teardown', t => {
  rimraf.sync('./test/output')
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_0/test_0')
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_0/test_1')
  t.end()
})

function fileLoc (options) {
  return `./test/data/files/${options.id}_${options.layer}/${options.key}/${options.name}.${options.format}`
}


test/xport.js
/* @flow */
'use strict'

const test = require('tape')
const rewire = require('rewire')
const xport = rewire('../src/jobs/xport')
const fs = require('fs')
const _ = require('highland')
const rimraf = require('rimraf')

test('Transform data when geojson already exists', t => {
  t.plan(2)
  const options = {
    id: 'f445febc447d4cb696e71ea7816d65d5',
    layer: 0,
    key: 'test',
    format: 'csv',
    name: 'test',
    path: './test/output'
  }
  xport(options, err => {
    t.error(err, 'No error')
    const rows = []
    _(fs.createReadStream(fileLoc(options)))
    .split()
    .compact()
    .each(r => rows.push(r))
    .done(() => t.equal(rows.length, 101), 'CSV has all expected features')
  })
})

test('Set up feature stream mock', t => {
  function createStream () {
    return _(fs.createReadStream('./test/fixtures/features.txt')).split().compact()
  }
  function exists (a, b, callback) { callback(false) }
  xport.__set__('koop.cache.createStream', createStream)
  xport.__set__('koop.files.exists', exists)
  t.end()
})

test('Transform data and save geojson when geojson does not exist', t => {
  t.plan(3)
  const options = {
    id: 'test',
    layer: 0,
    key: 'test',
    format: 'csv',
    name: 'test',
    path: './test/output'
  }
  xport(options, err => {
    t.error(err, 'No error')

    const path = fileLoc(options)
    const rows = []
    _(fs.createReadStream(path))
    .split()
    .compact()
    .each(r => rows.push(r))
    .done(() => t.equal(rows.length, 101), 'CSV has all expected features')

    const geojson = fs.readFileSync(path.replace('csv', 'geojson'))
    const features = JSON.parse(geojson).features
    t.equal(features.length, 100, 'GeoJSON has all expected features')
  })
})

test('Transform data directly into geojson', t => {
  t.plan(2)
  const options = {
    id: 'test',
    layer: 1,
    key: 'test',
    format: 'geojson',
    name: 'test',
    path: './test/output'
  }
  xport(options, err => {
    t.error(err, 'No error')
    const path = fileLoc(options)
    const geojson = fs.readFileSync(path)
    const features = JSON.parse(geojson).features
    t.equal(features.length, 100, 'GeoJSON has all expected features')
  })
})

test('Teardown', t => {
  rimraf.sync('./test/output')
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_0/test_0')
  rimraf.sync('./test/data/files/f445febc447d4cb696e71ea7816d65d5_0/test_1')
  t.end()
})

function fileLoc (options) {
  return `./test/data/files/${options.id}_${options.layer}/${options.key}/${options.name}.${options.format}`
}


src/jobs/xform.js

const path = require('path')
const config = require('config')

const createKoop = require('koop')
const koop = createKoop(config)

const log = koop.log
if (config.cache !== 'local') {
  const cache = require('koop-pgcache')
  koop.registerCache(cache)
}

const GeoXForm = require('geo-xform')
const _ = require('highland')

function xform (options, done) {
  options.filePath = path.join('files', `/${options.id}_${options.layer || 0}`, options.key)
  const fileName = `${options.name}.geojson`
  options.geojsonPath = path.join(options.filePath, fileName)
  if (options.format === 'geojson') return xformGeojson(options, done)
  koop.files.exists(options.filePath, fileName, exists => {
    if (exists) return xformOnly(options, done)
    xformAndSave(options, done)
  })
}

function xformOnly (options, done) {
  const fileName = options.geojsonPath.replace(/(?!\S+\.)geojson/, options.format)
  const output = koop.files.createWriteStream(fileName)
  koop.files.createReadStream(options.geojsonPath)
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => done(e))
  .pipe(GeoXForm.createStream(options.format, options))
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => done(e))
  .pipe(output)
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => done(e))
  .on('finish', () => done())
}

function xformAndSave (options, done) {
  const cacheStream = createCacheStream(options)
  const geojsonOut = koop.files.createWriteStream(options.geojsonPath)
  const fileName = `${options.name}.${options.format}`
  const output = koop.files.createWriteStream(path.join(options.filePath, fileName))

  // we can pipe the geojson straight to S3 while transformations are in progress
  // that way we can reuse the geojson rather than hitting the database again
  cacheStream
  .observe()
  .pipe(geojsonOut)
  .on('error', e => log.error(e))

  cacheStream
  .pipe(GeoXForm.createStream(options.format, options))
  .on('error', e => done(e))
  .on('log', l => log.apply(l.level, [l.message]))
  .pipe(output)
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => done(e))
  .on('finish', () => done())
}

function xformGeojson (options, done) {
  createCacheStream(options)
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => done(e))
  .pipe(koop.files.createWriteStream(options.geojsonPath))
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => done(e))
  .on('finish', () => done())
}

function createCacheStream (options) {
  const output = _()
  koop.Cache.createStream(options.table, options)
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => output.emit('error', e))
  .pipe(GeoXForm.GeoJSON.createStream())
  .on('log', l => log.apply(l.level, [l.message]))
  .on('error', e => output.emit('error', e))
  .pipe(output)
  return output
}

module.exports = xform


src/jobs/xport.js
/* @flow */
'use strict'
const path = require('path')
const config = require('config')

const createKoop = require('koop')
const koop = createKoop(config)

const log = koop.log
if (config.cache !== 'local') {
  const cache = require('koop-pgcache')
  koop.register(cache)
}

const GeoXForm = require('geo-xform')
const _ = require('highland')

function exportFile (options, done) {
  // TODO ya gotta fix this d00d
  if (!options.filePath) options.filePath = path.join('files', `/${options.id}_${options.layer || 0}`, options.key)
  const fileName = `${options.name}.geojson`
  options.geojsonPath = path.join(options.filePath, fileName)
  if (options.format === 'geojson') return exportGeojson(options, done)
  koop.files.exists(options.filePath, fileName, exists => {
    log.info('GeoJSON exists:', exists, options)
    if (exists) exportFromS3(options, done)
    else exportFromCache(options, done)
  })
}

function exportFromS3 (options, done) {
  const fileName = options.geojsonPath.replace(/(?!\S+\.)geojson/, options.format)
  const transform = GeoXForm.createStream(options.format, options)
  const output = koop.files.createWriteStream(fileName)

  koop.files.createReadStream(options.geojsonPath)
  .on('log', l => log[l.level](l.message))
  .on('error', e => done(e))
  .pipe(transform)
  .on('log', l => log[l.level](l.message))
  .on('error', e => done(e))
  .pipe(output)
  .on('log', l => log[l.level](l.message))
  .on('error', e => done(e))
  .on('finish', () => done())
}

function exportFromCache (options, done) {
  const cacheStream = createCacheStream(options)
  const fileName = `${options.name}.${options.format}`
  const output = koop.files.createWriteStream(path.join(options.filePath, fileName))

  cacheStream
  .pipe(GeoXForm.createStream(options.format, options))
  .on('error', e => done(e))
  .on('log', l => log[l.level](l.message))
	.on('data', () => console.log('data coming through the pipe'))
  .pipe(output)
  .on('log', l => log[l.level](l.message))
  .on('error', e => done(e))
  .on('finish', () => done())
}

function exportGeojson (options, done) {
  createCacheStream(options)
  .on('log', l => log[l.level](l.message))
  .on('error', e => done(e))
  .pipe(koop.files.createWriteStream(options.geojsonPath))
  .on('log', l => log[l.level](l.message))
  .on('error', e => done(e))
  .on('finish', () => done())
}

function createCacheStream (options) {
  const output = _()
  koop.cache.createStream(options.table, options)
  .on('log', l => log[l.level](l.message))
  .on('error', e => output.emit('error', e))
  .pipe(GeoXForm.GeoJSON.createStream())
  .on('log', l => log[l.level](l.message))
  .on('error', e => output.emit('error', e))
  .pipe(output)
  return output
}

module.exports = exportFile


src/plugins/retry.js
// If a job fails, retry it N times before finally placing it into the failed queue
// a port of some of the features in https://github.com/lantins/resque-retry

var os = require('os')
const crypto = require('crypto')

var retry = function (worker, func, queue, job, args, options) {
  var self = this

  if (!options.retryLimit) { options.retryLimit = 1 }
  if (!options.retryDelay) { options.retryDelay = (1000 * 5) }
  if (!options.backoffStrategy) { options.backoffStrategy = null }

  self.name = 'retry'
  self.worker = worker
  self.queue = queue
  self.func = func
  self.job = job
  self.args = args
  self.options = options

  if (self.worker.queueObject) {
    self.queueObject = self.worker.queueObject
  } else {
    self.queueObject = self.worker
  }
}

// //////////////////
// PLUGIN METHODS //
// //////////////////

retry.prototype.argsKey = function () {
  var self = this
  return crypto.createHash('sha1').update(self.args.join('-')).digest('hex')
}

retry.prototype.retryKey = function () {
  var self = this
  return self.queueObject.connection.key('resque-retry', self.func, self.argsKey()).replace(/\s/, '')
}

retry.prototype.failureKey = function () {
  var self = this
  return self.queueObject.connection.key('failure-resque-retry:' + self.func + ':' + self.argsKey()).replace(/\s/, '')
}

retry.prototype.maxDelay = function () {
  var self = this
  var maxDelay = self.options.retryDelay || 1
  if (Array.isArray(self.options.backoffStrategy)) {
    self.options.backoffStrategy.forEach(function (d) {
      if (d > maxDelay) { maxDelay = d }
    })
  }
  return maxDelay
}

retry.prototype.redis = function () {
  var self = this
  return self.queueObject.connection.redis
}

retry.prototype.attemptUp = function (callback) {
  var self = this
  var key = self.retryKey()
  self.redis().setnx(key, -1, function (error) {
    if (error) { return callback(error) }
    self.redis().incr(key, function (error, retryCount) {
      if (error) { return callback(error) }
      self.redis().expire(key, self.maxDelay(), function (error) {
        if (error) { return callback(error) }
        var remaning = self.options.retryLimit - retryCount - 1
        return callback(null, remaning)
      })
    })
  })
}

retry.prototype.saveLastError = function (callback) {
  var self = this
  var now = new Date()
  var failedAt = '' +
  now.getFullYear() + '/' +
  (('0' + (now.getMonth() + 1)).slice(-2)) + '/' +
  (('0' + now.getDate()).slice(-2)) + ' ' +
  (('0' + now.getHours()).slice(-2)) + ':' +
  (('0' + now.getMinutes()).slice(-2)) + ':' +
  (('0' + now.getSeconds()).slice(-2))

  var data = {
    failed_at: failedAt,
    payload: self.args,
    exception: String(self.worker.error),
    error: String(self.worker.error),
    backtrace: self.worker.error && self.worker.error.stack.split(os.EOL) || [],
    worker: self.func,
    queue: self.queue
  }

  self.redis().setex(self.failureKey(), self.maxDelay(), JSON.stringify(data), callback)
}

retry.prototype.cleanup = function (callback) {
  var self = this
  var key = self.retryKey()
  var failureKey = self.failureKey()
  self.redis().del(key, function (error) {
    if (error) { return callback(error) }
    self.redis().del(failureKey, function (error) {
      if (error) { return callback(error) }
      return callback()
    })
  })
}

retry.prototype.after_perform = function (callback) {
  var self = this

  if (!self.worker.error) return self.cleanup(callback)

  self.attemptUp(function (error, remaning) {
    if (error) { return callback(error) }
    self.saveLastError(function (error) {
      if (error) { return callback(error) }
      if (remaning <= 0) {
        self.cleanup(function (error) {
          if (error) { return callback(error) }
          return callback(self.worker.error, true)
        })
      } else {
        var nextTryDelay = self.options.retryDelay
        if (Array.isArray(self.options.backoffStrategy)) {
          var index = (self.options.retryLimit - remaning - 1)
          if (index > (self.options.backoffStrategy.length - 1)) {
            index = (self.options.backoffStrategy.length - 1)
          }
          nextTryDelay = self.options.backoffStrategy[index]
        }

        self.queueObject.enqueueIn(nextTryDelay, self.queue, self.func, self.args, function (error) {
          if (error) { return callback(error) }

          self.worker.emit('reEnqueue', self.queue, self.job, {
            delay: nextTryDelay,
            remaningAttempts: remaning,
            err: self.worker.error
          })

          self.redis().decr(self.queueObject.connection.key('stat', 'processed'))
          self.redis().decr(self.queueObject.connection.key('stat', 'processed', self.worker.name))

          self.redis().incr(self.queueObject.connection.key('stat', 'failed'))
          self.redis().incr(self.queueObject.connection.key('stat', 'failed', self.worker.name))

          callback(self.worker.error)
          delete self.worker.error
        })
      }
    })
  })
}

module.exports = retry


src/index.js
/* @flow */
'use strict'
const config = require('config')
const NR = require('node-resque')
const lodash = require('lodash')
const Redis = require('redis')
const connection = config.queue.connection || {}
connection.pkg = 'redis'
const redis = Redis.createClient(connection)
const Logger = require('koop-logger')
const log = new Logger(config)
const Queue = require('koop-queue')
const queue = new Queue()

const exportFile = require('./jobs/exportFile')
const copy = require('./jobs/copyFile')

let running

const jobs = {
  exportFile: {
    perform: (job, finish) => {
      const options = lodash.cloneDeep(job)
      running = exportFile(options, (err, result) => {
        finishJob(job, err, result, finish)
      })
    }
  },
  copyFile: {
    perform: (job, finish) => {
      const options = lodash.cloneDeep(job)
      copy(options, (err, result) => {
        finishJob(job, err, result, finish)
      })
    }
  }
}

function finishJob (job, error, result, callback) {
  clearInterval(heartbeat)
  if (error && shouldRetry(job, error)) {
    callback(null, {retried: true, error})
  } else {
    callback(error, result)
  }
}

function shouldRetry (job, err) {
  const retries = job.retries ? job.retries : 0
  // coerce maxRetries to integer
  const maxRetries = parseInt(job.maxRetries, 10)
  return retries < maxRetries && err.recommendRetry
}

const queues = config.queues || ['koop']

const worker = new NR.worker({connection, queues}, jobs) // eslint-disable-line
worker.connect(() => {
  worker.workerCleanup()
  worker.start()
})

let heartbeat

worker.on('job', (queue, job) => {
  fmtLog('Working on', job)
  publish('start', job)
  heartbeat = setInterval(() => {
    fmtLog('Working on', job)
    publish('progress', job)
  }, 5000)
})

// we have to force failures that need to be retried through this so we can override
// node-resque's success fail logic
worker.on('success', (queue, job, result) => {
  if (result && result.retried) {
    fmtLog('Job failed, retrying', job, result.error)
    worker.emit('retry', job)
  } else {
    fmtLog('Job finished', job)
    publish('finish', job)
  }
})

worker.on('retry', job => {
  fmtLog('Reenqueing:', job)
  publish('retry', job)
  // unfreeze the job object
  job = JSON.parse(JSON.stringify(job))
  if (!job.args[0].retries) {
    job.args[0].retries = 1
  } else {
    job.args[0].retries++
  }
  job.args.delay = 30000
  queue.enqueue(job.class, job.args)
})

worker.on('failure', (queue, job, error) => {
  if (!error.domainThrow) fmtLog('Job failed', job, error)
  publish('fail', job, error)
  // if we've landed here an error was caught in domain
  // rather than potentially leak resources we just shut down
  if (error.domainThrown) {
    fmtLog('Worker failed in unknown state, shutting down in 5 seconds', job, error)
    worker.end()
    setTimeout(() => process.exit(1), 5000)
  }
})

worker.on('error', (queue, job, error) => {
  clearInterval(heartbeat)
  fmtLog('Worker emitted error', job, error)
})

worker.on('start', () => log.info('Worker started'))

worker.on('end', () => {
  log.info('Worker shutting down')
  redis.quit()
  queue.shutdown()
  clearInterval(heartbeat)
  clearInterval(checker)
})

function publish (status, job, error) {
  let id
  if (job && job.args && job.args.length) {
    job = job.args[0]
    id = job.job_id
  }
  let errorReport
  if (error) {
    errorReport = {
      message: error.message,
      code: error.code,
      time: error.time
    }
  }
  const info = JSON.stringify({id, status, job, errorReport})
  redis.publish('jobs', info)
}

function fmtLog (msg, job, error) {
  if (error) log.error(msg, fmtJob(job), fmtError(error))
  else log.info(msg, fmtJob(job))
}

function fmtJob (job) {
  var logable
  if (job && job.args && job.args.length) logable = job.args[0]
  else logable = job
  return lodash.omit(logable, 'metadata')
}

function fmtError (error) {
  if (!error || !error.stack) return ''
  return error.stack.match(/.+\n.+[^\n]/)[0]
}

process
.on('SIGINT', () => {
  if (running) running.abort('SIGTERM')
  worker.end(() => process.exit())
})
.on('SIGTERM', () => {
  if (running) running.abort('SIGTERM')
  worker.end(() => process.exit())
})
.on('unhandledRejection', (error, p) => {
  log.error('Worker failed with unhandled rejection, shutting down in 5 seconds', error)
  worker.end()
  setTimeout(() => process.exit(1), 5000)
})

const checker = setInterval(() => {
  if (running && running.abort) {
    const tooLong = 60 * 60 * 1000
    if (Date.now() - running.start > tooLong) running.abort('Job aborted, took too long')
  }
}, 5 * 60 * 1000)

module.exports = worker
